File: ASD-labs-main/Lab9/t09_04_e2321.py
# Завдання 9.4 by  Янголь Ярослав / Комп. мех / 2 курс
def QuickSort(arr, a, b):
    if a >= b:
        return

    pivot = arr[a + (b - a) // 2]
    l = a
    r = b

    while True:
        while arr[l] < pivot:
            l += 1
        while arr[r] > pivot:
            r -= 1
        if l >= r:
            break

        arr[l], arr[r] = arr[r], arr[l]
        l += 1
        r -= 1

    QuickSort(arr, a, r)
    QuickSort(arr, r + 1, b)


if __name__ == '__main__':
    n = int(input())
    nums = list(map(int, input().split()))
    QuickSort(nums, 0, n - 1)
    for i in range(n):
        print(nums[i], end=' ')
    print()

File: ASD-labs-main/Lab9/t09_04_e2321.cpp
// Завдання 8.7 by  Янголь Ярослав / Комп. мех / 2 курс
#include <iostream>

using namespace std;


// Quick Sort
int QuickSort(int arr[], int a, int b) {

    if (a >= b) return 0;

    int pivot = arr[a + (b - a) / 2];
    int l = a;
    int r = b;

    while (1){
        while (arr[l] <= pivot){
            l++;
        }
        while (arr[r] >= pivot){
            r--;
        }
        if (l >= r) break;
        int temp = arr[l];
        arr[l] = arr[r];
        arr[r] = temp;
        l++; r--;
    }

    QuickSort(arr, a, r);
    QuickSort(arr, r + 1, b);
    return 0;
}


int main(){

    int n;
    cin >> n;
    int nums[n];
    for (int i = 0; i < n; i++){
        cin >> nums[i];
    }
    QuickSort(nums, 0, n - 1);
    for (int i = 0; i < n; i++){
        cout << nums[i] << " ";
    }
    cout << endl;

    return 0;

}
